### PR TITLE
fix(api): topic flows should use external subscriber id (user's custom id)

### PR DESCRIPTION
**Pull request by p-fernandez**
### What change does this PR introduce? 

 
Relation between topics and the subscribers will be based on the external subscriber ID. Also we check in the database before adding the subscribers that they belong to the enviroment and organization of the user.

### Why was this change needed?

I have done all the development of the Topics feature assuming the customers would use our subscriber._id as the identifier that they would send and use for adding and removing subscribers to a topic. But the problem is that the users use their own subscriberId that is represented in our schema as subscriber.subscriberId to operate like when triggering an event, is that the subscriber ID used in the field to of the payload.

This task is to redo the Topics feature codebase to support that in the expected flow and fix my wrong assumption.

### CODE REVIEW

File: apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.use-case.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      externalSubscriberIds: command.externalSubscriberIds,
+    };
+  }
+
+  private mapFromEntity(entity: SubscriberEntity): SubscriberDto {
+    const { _id, createdAt, updatedAt, ...rest } = entity;
+
+    return {
+      ...rest,
+      _id,
```
Comment (p-fernandez) [inline apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.use-case.ts:32]: The conditional ids in the entities...

File: apps/api/src/app/topics/dtos/add-subscribers.dto.ts [hunk 2]
```diff
@@ -1,13 +1,13 @@
 import { ApiProperty } from '@nestjs/swagger';
 import { IsArray, IsDefined } from 'class-validator';
 
-import { SubscriberId } from '../types';
+import { ExternalSubscriberId } from '../types';
```
Comment (p-fernandez) [inline apps/api/src/app/topics/dtos/add-subscribers.dto.ts:4]: All the confusion came from using the same term both for our internal and the external subscriber ID. Hope the types help to clarify the use when needed.

File: apps/api/src/app/topics/e2e/remove-subscribers.e2e.ts [hunk 3]
```diff
@@ -71,11 +71,33 @@ describe('Remove subscribers to topic - /topics/:topicId/subscribers/removal (PO
     expect(getResponseTopic._organizationId).to.eql(session.organization._id);
     expect(getResponseTopic.key).to.eql(topicKey);
     expect(getResponseTopic.name).to.eql(topicName);
-    expect(getResponseTopic.subscribers).to.eql([secondSubscriber._id, thirdSubscriber._id]);
+    expect(getResponseTopic.subscribers).to.eql([secondSubscriber.subscriberId, thirdSubscriber.subscriberId]);
+  });
+
+  it('should not remove subscriber from topic if it does not exist', async () => {
```
Comment (p-fernandez) [inline apps/api/src/app/topics/e2e/remove-subscribers.e2e.ts:77]: For the removal the check is not much needed as result of this passing test.

File: apps/api/src/app/topics/use-cases/add-subscribers/add-subscribers.use-case.ts [hunk 4]
```diff
@@ -1,20 +1,64 @@
...
+  constructor(
+    private searchByExternalSubscriberIds: SearchByExternalSubscriberIds,
+    private topicSubscribersRepository: TopicSubscribersRepository
+  ) {}
 
   async execute(command: AddSubscribersCommand) {
-    const entity = this.mapToEntity(command);
+    const filteredSubscribersCommand = await this.filterExistingSubscribers(command);
 
+    const entity = this.mapToEntity(filteredSubscribersCommand);
```
Comment (p-fernandez) [inline apps/api/src/app/topics/use-cases/add-subscribers/add-subscribers.use-case.ts:20]: Yet to decide how we will inform in the API which subscribers we could save and which ones didn't. To be done in a different PR.

File: libs/dal/src/repositories/subscriber/subscriber.schema.ts [hunk 5]
```diff
@@ -27,8 +27,8 @@ const subscriberSchema = new Schema(
   schemaOptions
 );
 
-subscriberSchema.index({ _environmentId: 1, userId: 1 });
```
Comment (p-fernandez) [inline libs/dal/src/repositories/subscriber/subscriber.schema.ts:30]: We have no userId in this schema... Unless I am missing anything this complex index is not needed.

File: libs/dal/src/repositories/topic/types.ts [hunk 6]
```diff
@@ -1,9 +1,9 @@
 import { Types } from 'mongoose';
 
-export { SubscriberId } from '../subscriber/types';
+export { EnvironmentId } from '../environment';
+export { ExternalSubscriberId } from '../subscriber';
+export { OrganizationId } from '../organization';
```
Comment (p-fernandez) [inline libs/dal/src/repositories/topic/types.ts:5]: Tidying up shareable stuff.

File: apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.spec.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    session = new UserSession();
+    await session.initialize();
+
+    useCase = moduleRef.get<SearchByExternalSubscriberIds>(SearchByExternalSubscriberIds);
+    subscribersService = new SubscribersService(session.organization._id, session.environment._id);
+    firstSubscriber = await subscribersService.createSubscriber();
+    secondSubscriber = await subscribersService.createSubscriber();
+  });
+
+  it('should seach and find the subscribers by the external subscriber ids', async () => {
```
Comment (scopsy) [inline apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.spec.ts:34]:
```suggestion
  it('should search and find the subscribers by the external subscriber ids', async () => {
```

File: apps/api/src/app/topics/dtos/add-subscribers.dto.ts (hunk 2 above)
Comment (scopsy) [reply apps/api/src/app/topics/dtos/add-subscribers.dto.ts:4]: I think it does. Good call.

File: libs/dal/src/repositories/subscriber/subscriber.entity.ts [hunk 8]
```diff
@@ -1,9 +1,10 @@
 import { ChatProviderIdEnum, PushProviderIdEnum } from '@novu/shared';
 
-import { SubscriberId } from './types';
+import { ExternalSubscriberId } from './types';
 
 export class SubscriberEntity {
-  _id?: SubscriberId;
+  // TODO: Use SubscriberId. Means lot of changes across whole codebase. Cool down.
```
Comment (scopsy) [inline libs/dal/src/repositories/subscriber/subscriber.entity.ts:6]: Mind creating a Linear ticket for this?

File: libs/dal/src/repositories/subscriber/subscriber.schema.ts (hunk 5 above)
Comment (scopsy) [reply libs/dal/src/repositories/subscriber/subscriber.schema.ts:30]: Good catch

File: libs/dal/src/repositories/subscriber/subscriber.schema.ts [hunk 9]
```diff
@@ -27,8 +27,8 @@ const subscriberSchema = new Schema(
...
 subscriberSchema.index({ _environmentId: 1, subscriberId: 1 });
+subscriberSchema.index({ _environmentId: 1, _organizationId: 1, subscriberId: 1 });
```
Comment (scopsy) [inline libs/dal/src/repositories/subscriber/subscriber.schema.ts:31]:
```suggestion
```

I actually think that this one is not needed, since the above index is enough specific to find the query. Mongo is smart enough to use the above one even you perform a search with organizationId.

File: libs/dal/src/repositories/topic/topic-subscribers.schema.ts [hunk 10]
```diff
@@ -31,7 +31,7 @@ const topicSubscribersSchema = new Schema(
       index: true,
       required: true,
     },
-    subscribers: [{ type: Schema.Types.ObjectId, ref: 'Subscriber', required: true }],
+    subscribers: [{ type: Schema.Types.String, field: 'subscriberId', ref: 'Subscriber', required: true }],
```
[the remark this answers is not in the document]
Comment (scopsy) [reply libs/dal/src/repositories/topic/topic-subscribers.schema.ts:34]: Might have missed it from previous PR's, but why not doing an object mapping with a _subcriberId on each topic subscriber entity instead of having all subscribers in the array here? I'm afraid we might be limited by mongo object size for very large topics.

So basically meaning to follow a more SQL-ish pattern for topic subscribers collection. So each subscriber will have his own document for the subscription. It might be better because that we will be able to save metadata for the topic subscription for individual subscriber, for example specific user preferences or other metadata. 

Or am I missing something here?

[the remark this answers is not in the document]
Comment (p-fernandez) [reply libs/dal/src/repositories/topic/topic-subscribers.schema.ts:34]: Ok. Let's change the structure of the document.

File: apps/api/src/app/topics/use-cases/index.ts [hunk 11]
```diff
@@ -16,7 +15,6 @@ export * from './remove-subscribers';
 export const USE_CASES = [
   AddSubscribersUseCase,
   CreateTopicUseCase,
-  CreateTopicSubscribersUseCase,
```
Comment (p-fernandez) [inline apps/api/src/app/topics/use-cases/index.ts:19]: With the change of schema and domain is not needed anymore.

File: libs/dal/src/repositories/base-repository.ts [hunk 12]
```diff
@@ -88,12 +86,14 @@ export class BaseRepository<T_Query, T_Response> {
...
-      });
-    });
+  async createMany(data: T_Query[]): Promise<InsertManyResult<Document>> {
+    const saved = await this.MongooseModel.collection.insertMany(data);
+
+    return {
+      acknowledged: saved.acknowledged,
+      insertedCount: saved.insertedCount,
+      insertedIds: saved.insertedIds,
+    };
```
Comment (p-fernandez) [inline libs/dal/src/repositories/base-repository.ts:96]: The implementation wasn't working and during one of my trials to see how to upsert I implemented it. I am not using it at the end, I can leave it or remove it.

File: libs/dal/src/repositories/base-repository.ts [hunk 13]
```diff
@@ -113,6 +113,12 @@ export class BaseRepository<T_Query, T_Response> {
     };
   }
 
+  async upsertMany(data: T_Query[]) {
+    const promises = data.map((entry) => this.MongooseModel.findOneAndUpdate(entry, entry, { upsert: true }));
+
+    return await Promise.all(promises);
+  }
```
Comment (p-fernandez) [inline libs/dal/src/repositories/base-repository.ts:120]: The other option recommend was to use `bulkWrite` but I found it harder to make it work with all the changes I had to do. Maybe we can review this and its performance in the near future.

Comment (scopsy) [reply libs/dal/src/repositories/base-repository.ts:120]: This is totally fine for now, can re evelaute later.

File: apps/api/src/app/topics/use-cases/filter-topics/filter-topics.use-case.ts [hunk 14]
```diff
@@ -35,20 +19,40 @@ export class FilterTopicsUseCase {
...
+    command: FilterTopicsCommand
+  ): Pick<TopicEntity, '_environmentId' | 'key' | '_organizationId'> {
+    return {
+      _environmentId: TopicRepository.convertStringToObjectId(command.environmentId),
+      _organizationId: TopicRepository.convertStringToObjectId(command.organizationId),
+      ...(command.key && { key: command.key }),
+    };
+  }
+
+  private mapFromEntityToDto(topic: TopicEntity & { subscribers: ExternalSubscriberId[] }): TopicDto {
```
Comment (LetItRock) [inline apps/api/src/app/topics/use-cases/filter-topics/filter-topics.use-case.ts:50]: the returned type should be a partial from the `TopicDto` as we don't return subscribers

File: apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.use-case.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+    const entities = await this.subscriberRepository.searchByExternalSubscriberIds(entity);
+
+    return entities.map(this.mapFromEntity);
+  }
+
+  private mapToEntity(command: SearchByExternalSubscriberIdsCommand): IExternalSubscribersEntity {
+    return {
+      _environmentId: command.environmentId,
+      _organizationId: command.organizationId,
+      externalSubscriberIds: command.externalSubscriberIds,
```
Comment (LetItRock) [inline apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.use-case.ts:23]: I'm wondering about two things here:
- why sometimes we do convert with `convertStringToObjectId` and sometimes not
- can we write a util that will do this conversion for us automatically, like `mapToEntity<Dto, Entity>(dto, ['environmentId', 'organizationId'])` like the second arg is an array of fields that should have names converted to `_name`

File: apps/api/src/app/topics/use-cases/add-subscribers/add-subscribers.use-case.ts [hunk 16]
```diff
@@ -1,27 +1,68 @@
...
+  private async filterExistingSubscribers(command: AddSubscribersCommand): Promise<SubscriberDto[]> {
+    const { environmentId, organizationId, subscribers } = command;
+    const searchByExternalSubscriberIdsCommand = SearchByExternalSubscriberIdsCommand.create({
+      environmentId: command.environmentId,
+      organizationId: command.organizationId,
+      externalSubscriberIds: subscribers,
+    });
+    const existingSubscribers = await this.searchByExternalSubscriberIds.execute(searchByExternalSubscriberIdsCommand);
+
+    return this.getIntersection(subscribers, existingSubscribers);
```
[the remark this answers is not in the document]
Comment (p-fernandez) [reply apps/api/src/app/topics/use-cases/add-subscribers/add-subscribers.use-case.ts:37]: You are right. We are not passing the list of unique subscribers, we pass the value from the request of the user as it is to check for integrity. This method had a different purpose but when I had to change the schema I forgot to review it properly as just shaped the returned list (the one with TopicSubscriberEntity[]).
The only thing is that `getIntersection` method would be needed to achieve this task: https://linear.app/novu/issue/NV-1364/feature-provide-proper-api-response-adding-subscribers
Though right now is incomplete. So I'd kindly ask to leave it like this for now.

File: apps/api/src/app/topics/use-cases/filter-topics/filter-topics.use-case.ts (hunk 14 above)
Comment (p-fernandez) [reply apps/api/src/app/topics/use-cases/filter-topics/filter-topics.use-case.ts:50]: No, the returned type is right. We are returning subscribers from:
https://github.com/novuhq/novu/blob/58bb5b321274ce7258d2fec7ba3a3203eda8d920/libs/dal/src/repositories/topic/topic.repository.ts#L54
We are returning the `Promise<TopicEntity & { subscribers: ExternalSubscriberId[] }> ` that then mapped to TopicDto that includes the `subscribers` field.
Check that we are doing in that repo method a lookup with a projection to provide the subscribers:
https://github.com/novuhq/novu/blob/58bb5b321274ce7258d2fec7ba3a3203eda8d920/libs/dal/src/repositories/topic/topic.repository.ts#L18

File: apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.use-case.ts (hunk 15 above)
Comment (p-fernandez) [reply apps/api/src/app/subscribers/usecases/search-by-external-subscriber-ids/search-by-external-subscriber-ids.use-case.ts:23]: About the first maybe a lack of consistency on my side due the huge amount of this work. I tried to apply this rule of thumb:
- for commands and dtos use a string therefore converting using `convertObjectIdToString`.
- for commands or queries to the DAL use already the ObjectId therefore converting using `convertStringToObjectId`

Regarding the second point I not fully totally understand but I already added a point to the Tech forum meeting regarding this functionalities that I added and if we wanted to expand them further, reshape them or use functionalities like the one you described. Basically I came with this quick implementation to handle transformations to DAL avoiding to creep a Mongoose import dependency into the API. That was my goal.

File: apps/api/src/app/topics/use-cases/filter-topics/filter-topics.use-case.ts (hunk 14 above)
Comment (LetItRock) [reply apps/api/src/app/topics/use-cases/filter-topics/filter-topics.use-case.ts:50]: ahh damn I've missed that as usual :D

**Review by LetItRock — APPROVED** [review]
